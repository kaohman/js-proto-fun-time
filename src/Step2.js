import React, {Component} from 'react';

class Step2 extends Component {
  constructor(props) {
    super(props);
    this.state = {
      answerInput: '',
      correctAnswer: ''
    }
  }

  getAnswerInput = (event) => {
    this.setState({
      answerInput: event.target.value,
      correctAnswer: ''
    });
  }

  checkAnswer = (event) => {
    event.preventDefault();
    let correctAnswer = this.state.answerInput.toLowerCase() === this.props.correctMethod.toLowerCase() ? 
      'Correct! Click Next Step to continue.' : 
      'Sorry that is incorrect, please try again.';
    this.setState({
      correctAnswer: correctAnswer
    });
  }

  render() {
    let { currentStep, incrementStep } = this.props;

    return (
      <div className='step-containers'>
        <div>
          <h3>Step 2: Pick Your Method</h3>
          <p>Which prototype method is best to use for this problem (ex. sort)?</p>
          <label>Answer:
          <input onChange={this.getAnswerInput}></input>
          </label>
          {
            currentStep === 2 && <h4>{this.state.correctAnswer}</h4>
          }
        </div>
        {
          currentStep === 2 && (
            this.state.correctAnswer === 'Correct! Click Next Step to continue.' ? 
              <button onClick={incrementStep}>Next Step</button> :
              <button onClick={this.checkAnswer}>Check Answer</button>
            )
        }
      </div>
    )
  }
}

export default Step2;