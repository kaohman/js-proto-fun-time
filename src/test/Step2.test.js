import React from 'react';
import ReactDOM from 'react-dom';
import Step2 from '../Step2';
import { shallow } from 'enzyme';

const incrementStepMock = jest.fn();
const correctMethodMock = 'map';
const currentStepMock = 2;
const changeEventMock = { target: { value: 'map'} };
const clickEventMock = { preventDefault: () => { } };

describe('Step2', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Step2 
        incrementStep={incrementStepMock}
        correctMethod={correctMethodMock}
        currentStep={currentStepMock}
      />
    );
  });

  it('should match the snapshot with all data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({ 
      answerInput: '',
      correctAnswer: ''
     });
  });

  it('should get answer input as the user types in the input box', () => {
    expect(wrapper.state('answerInput')).toEqual('');
    wrapper.find('input').simulate('change', changeEventMock);
    expect(wrapper.state('answerInput')).toEqual('map');
  });

  it('should check the user answer against the correct answer when the user clicks a button and tell the user if the answer is correct', () => {
    expect(wrapper.state('correctAnswer')).toEqual('');
    wrapper.find('input').simulate('change', changeEventMock);
    wrapper.find('#check-answer-button-2').simulate('click', clickEventMock);
    expect(wrapper.state('correctAnswer')).toEqual('Correct! Click Next Step to continue.');
  });

  it('should tell the user if the answer is incorrect', () => {
    expect(wrapper.state('correctAnswer')).toEqual('');
    wrapper.setState({ answerInput: 'forEach' });
    wrapper.find('#check-answer-button-2').simulate('click', clickEventMock);
    expect(wrapper.state('correctAnswer')).toEqual('Sorry that is incorrect, please try again.');
  });

  it('should allow the user to click the Next Step button if their answer was correct', () => {
    wrapper.setState({ correctAnswer: 'Correct! Click Next Step to continue.' });
    wrapper.find('#next-step-button-2').simulate('click');
    expect(incrementStepMock).toBeCalled();
  });  

});