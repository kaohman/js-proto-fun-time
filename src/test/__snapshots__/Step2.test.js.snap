// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step2 should match the snapshot with all data passed in 1`] = `
<div
  className="step-containers"
  id="step-2-container"
>
  <div>
    <h3>
      Step 2: Pick Your Method
    </h3>
    <p>
      Which prototype method is best to use for this problem (ex. sort)?
    </p>
    <label>
      Answer:
      <input
        onChange={[Function]}
      />
    </label>
    <h4 />
  </div>
  <button
    id="check-answer-button-2"
    onClick={[Function]}
  >
    Check Answer
  </button>
</div>
`;
